gear/gear/auth.py

```diff
@@ -8,96 +8,101 @@
 
 log = logging.getLogger('gear.auth')
 
+deploy_config = get_deploy_config()
 
-def _authenticated_users_only(rest, redirect):
-    deploy_config = get_deploy_config()
+
+async def _userdata_from_session_id(session_id):
+    headers = {'Authorization': f'Bearer {session_id}'}
+    try:
+        async with aiohttp.ClientSession(
+                raise_for_status=True, timeout=aiohttp.ClientTimeout(total=60)) as session:
+            async with session.get(deploy_config.url('auth', '/api/v1alpha/userinfo'),
+                                   headers=headers) as resp:
+                return await resp.json()
+    except Exception:  # pylint: disable=broad-except
+        log.exception('getting userinfo')
+        return None
+
+
+async def userdata_from_web_request(request):
+    session = await aiohttp_session.get_session(request)
+    if 'session_id' not in session:
+        return None
+    return await _userdata_from_session_id(session['session_id'])
+
+
+async def userdata_from_rest_request(request):
+    if 'Authorization' not in request.headers:
+        return None
+    auth_header = request.headers['Authorization']
+    if not auth_header.startswith('Bearer '):
+        return None
+    return await _userdata_from_session_id(auth_header[7:])
+
+
+def rest_authenticated_users_only(fun):
+    async def wrapped(request, *args, **kwargs):
+        userdata = await userdata_from_rest_request(request)
+        if not userdata:
+            raise web.HTTPUnauthorized()
+        return await fun(request, userdata, *args, **kwargs)
+    return wrapped
+
+
+def _web_unauthorized(request, redirect):
+    if not redirect:
+        raise web.HTTPUnauthorized()
+
+    login_url = deploy_config.external_url('auth', '/login')
+
+    # request.url is yarl.URL
```

Comment:
what is yarl?

Reply:
Where is yarl imported?

Reply:
As an aside, type hints would be awesome (In the below, what is the type of service_namespace? hard to tell from glancing at the consumer):

<img width="491" alt="Screen Shot 2019-09-25 at 6 35 07 PM" src="https://user-images.githubusercontent.com/5543229/65644772-35753000-dfc3-11e9-9e52-d716e2cb8726.png">

Reply:
Yarl is not imported by us, but by aiohttp.  I guess this should say `request.url is A yarl.URL`.